packages/kotti-input-number/src/InputNumber.vue

```diff
@@ -137,70 +148,91 @@ export default {
+					throw new RangeError(
+						`KtInputNumber: encounted an out-of-range number "${newNumber}"`,
+					)
 
-				if (shouldUpdate) this.setValue(newString)
+				if (!isStepMultiple({ min, step, value: truncatedNumber }))
+					throw new Error(
+						`KtInputNumber: encounted a value "${newNumber}" that doesn't fit ((min + k * step): where k is an integer)`,
+					)
+
+				const shouldUpdate = oldNumber !== truncatedNumber
+				if (shouldUpdate) this.setInternalStringValue(newString)
 			},
 			immediate: true,
 		},
 	},
 	methods: {
 		decrementValue() {
 			if (!this.isDecrementEnabled) return
-			if (this.currentValueNumber === null) {
-				const defaultString = this.min === null ? '0' : this.min + ''
-				this.setValue(defaultString)
+			if (this.value === null) {
+				const defaultNumber = this.min || 0
+				this.emitInput(defaultNumber)
 				return
 			}
-
-			this.setValue(toString(this.currentValueNumber - this.step))
+			this.emitInput(this.value - this.step)
+		},
+		emitInput(value) {
+			this.$emit('input', value)
 		},
 		handleBlur() {
-			this.setValue(toString(this.currentValueNumber))
+			this.setInternalStringValue(toString(this.value))
 		},
 		handleInput(value) {
-			const { max, min } = this
+			const { max, min, step } = this
 
 			const valueWithoutLeadingZeroes = value.replace(
 				LEADING_ZEROES_REGEX,
 				'$1',
 			)
+			const nextNumber = toNumber(valueWithoutLeadingZeroes)
 
 			const isTypedNumberValid =
 				VALID_REGEX.test(valueWithoutLeadingZeroes) &&
-				(STRINGS_THAT_ARE_TREATED_AS_NULL.includes(valueWithoutLeadingZeroes) ||
-					isInRange({ max, min, value: toNumber(valueWithoutLeadingZeroes) }))
-
-			if (isTypedNumberValid) this.setValue(valueWithoutLeadingZeroes)
-			else this.hasFormError = true
-
+				isStepMultiple({
+					min,
+					step,
+					value: nextNumber,
+				}) &&
+				isInRange({
+					max,
+					min,
+					offset: 0,
+					value: nextNumber,
+				})
+
+			if (isTypedNumberValid) {
+				const shouldEmit = nextNumber !== this.value
+				if (shouldEmit) this.emitInput(nextNumber)
+				else this.setInternalStringValue(valueWithoutLeadingZeroes)
+			} else {
+				this.hasFormError = true
+				this.$emit('error', valueWithoutLeadingZeroes)
+			}
 			// vue doesn't support controlled input fields without re-rendering
 			// therefore, in case nothing changed, we need to re-render here
 			this.$forceUpdate()
 		},
 		incrementValue() {
 			if (!this.isIncrementEnabled) return
-			if (this.currentValueNumber === null) {
-				const defaultString = this.min === null ? '0' : this.min + ''
-				this.setValue(defaultString)
+			if (this.value === null) {
+				const defaultNumber = this.min || 0
+				this.emitInput(defaultNumber)
 				return
 			}
 
-			this.setValue(toString(this.currentValueNumber + this.step))
+			this.emitInput(this.value + this.step)
 		},
-		setValue(newValue) {
-			const oldNumber = this.currentValueNumber
+		setInternalStringValue(value) {
```

Comment:
internal setter, that shouldn't emit to avoid the above-mentioned issue with the `12.`